### ISSUE TITLE
P4 table properties

### ISSUE THREAD

**Issue by jafingerhut** [jafingerhut/p4-namespaces#6]
Should table properties be importable from a namespace?

If so, first they should probably be declarable in some explicit way.  There is no such explicit way to declare them defined in the P4_16 language spec that I am aware of.  For example, we could say that they should be declared similar to match_kind, perhaps with a syntax like this:

```
table_properties {
    key,
    actions,
    default_action,
    entries,
    size
}
```

Then the answers to how they should be named after being imported should hopefully, for consistency's sake, be the same as how they are done for match_kind, however that is (see https://github.com/jafingerhut/p4-namespaces/issues/5)

Table properties defined in the P4_16 language specification include: key, actions, default_action, entries, size (see https://p4.org/p4-spec/docs/P4-16-v1.2.2.html#sec-table-props)

Table properties added in v1model: counter, meter, implementation, idle_timeout, there could be more but I do not know where to find a complete documented list without reading all of the p4c v1model back end code.

Table properties added by PSA v1.1: psa_direct_counter, psa_direct_meter, psa_implementation, psa_empty_group_action, psa_idle_timeout  (see https://p4.org/p4-spec/docs/PSA-v1.1.0.html#sec-psa-table-properties)

**Comment by mbudiu-vmw** [comment]
Indeed, in the languages I can think of only declarations can be imported, so there is nothing you can do about table properties.

**Comment by jafingerhut** [comment]
This is a design decision.  We _could_ choose to say "table properties must be declared before they can be used, and they can be imported between modules/namespaces".  I don't understand the viewpoint that this is impossible.  It might be undesirable.  It might be difficult to implement starting from the implementation we have now.  It isn't a fundamental law of the universe that we cannot do anything about it, though.
